go-controller/pkg/ovn/egressip.go

```diff
@@ -0,0 +1,542 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	goovn "github.com/ebay/go-ovn"
+	egressipv1 "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	kapi "k8s.io/api/core/v1"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+	utilnet "k8s.io/utils/net"
+)
+
+type modeEgressIP interface {
+	addPod(eIP *egressipv1.EgressIP, pod *kapi.Pod) error
+	deletePod(eIP *egressipv1.EgressIP, pod *kapi.Pod) error
+	needsRetry(pod *kapi.Pod) bool
+}
+
+func (oc *Controller) addEgressIP(eIP *egressipv1.EgressIP) error {
+	// If the status is set at this point, then we know it's valid from syncEgressIP and we have no assignment to do.
+	// Just initialize all watchers (which should not re-create any already existing items in the OVN DB)
+	if len(eIP.Status) == 0 {
+		assignments, err := oc.assignEgressIPs(eIP)
+		if err != nil {
+			klog.Errorf("Unable to assign egress IP: %s, error: %v", eIP.Name, err)
+		}
+		eIP.Status = assignments
+	}
+
+	oc.egressIPNamespaceHandlerMutex.Lock()
+	defer oc.egressIPNamespaceHandlerMutex.Unlock()
+
+	h, err := oc.watchFactory.AddFilteredNamespaceHandler("", &eIP.Spec.NamespaceSelector,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
```

Comment:
why do we need to watch namespaces? AFAICT we only care about pods in a namespace. So we just watch for pods in a specific namespace, with or without a user defined pod selector.

Reply:
No, since https://issues.redhat.com/browse/RFE-180 was accepted we are expected to have the egress IP feature be applied to either a set of namespaces / set of pods in a namespace(s). A bit like network policies.